### PR TITLE
[Kinder] Remove 1.16 workflows

### PR DESCRIPTION
**Pull request by Akhilesh2412**
This CL removes v1.16 kubeadm CI tests. These are E2E tests
running periodically which are no longer needed in v1.20.

Signed-off-by: Akhilesh Kommireddy <akhileshkommireddy2412@gmail.com>

### CODE REVIEW

**Comment by fabriziopandini** [comment]
@Akhilesh2412 please follow instructions to sign the CLA

**Comment by fabriziopandini** [comment]
/approve 
If I'm not wrong this PR should wait for other tasks in https://github.com/kubernetes/kubeadm/issues/2277, but I leave the final word to @neolit123 who is coordinating this effort

**Comment by neolit123** [comment]
this PR should wait for the 1.19.1 release:
>  [TODO] Cleanup E2E jobs for 1.x-3 (requires two PRs)
https://github.com/kubernetes/kubeadm/issues/2277

it should also remove the e2e tests for 1.16 under /kinder/ci/workflows.

File: kinder/pkg/cluster/manager/actions/kubeadm-init.go [hunk 1]
```diff
@@ -40,9 +40,9 @@ import (
 func KubeadmInit(c *status.Cluster, usePhases, kubeDNS, automaticCopyCerts bool, kustomizeDir, patchesDir string, wait time.Duration, vLevel int) (err error) {
 	cp1 := c.BootstrapControlPlane()
 
-	// fail fast if required to use kustomize and kubeadm less than v1.16
-	if kustomizeDir != "" && cp1.MustKubeadmVersion().LessThan(constants.V1_16) {
-		return errors.New("--kustomize-dir can't be used with kubeadm older than v1.16")
+	// fail fast if required to use kustomize and kubeadm less than v1.17
+	if kustomizeDir != "" && cp1.MustKubeadmVersion().LessThan(constants.V1_17) {
+		return errors.New("--kustomize-dir can't be used with kubeadm older than v1.17")
```
Comment (neolit123) [inline kinder/pkg/cluster/manager/actions/kubeadm-init.go:45]: once the 1_16 constant is removed we can remove the whole check with the assumption that 1.16 test jobs will not run.

File: kinder/pkg/cluster/manager/actions/kubeadm-join.go [hunk 2]
```diff
@@ -43,9 +43,9 @@ func joinControlPlanes(c *status.Cluster, usePhases, automaticCopyCerts bool, di
 	cpX := []*status.Node{c.BootstrapControlPlane()}
 
 	for _, cp2 := range c.SecondaryControlPlanes().EligibleForActions() {
-		// fail fast if required to use kustomize and kubeadm less than v1.16
-		if kustomizeDir != "" && cp2.MustKubeadmVersion().LessThan(constants.V1_16) {
-			return errors.New("--kustomize-dir can't be used with kubeadm older than v1.16")
+		// fail fast if required to use kustomize and kubeadm less than v1.17
+		if kustomizeDir != "" && cp2.MustKubeadmVersion().LessThan(constants.V1_17) {
+			return errors.New("--kustomize-dir can't be used with kubeadm older than v1.17")
```
Comment (neolit123) [inline kinder/pkg/cluster/manager/actions/kubeadm-join.go:48]: ditto.

File: kinder/pkg/cluster/manager/actions/kubeadm-upgrade.go [hunk 3]
```diff
@@ -45,9 +45,9 @@ func KubeadmUpgrade(c *status.Cluster, upgradeVersion *K8sVersion.Version, kusto
 
 	for _, n := range nodeList {
 
-		// fail fast if required to use kustomize and kubeadm less than v1.16
-		if kustomizeDir != "" && n.MustKubeadmVersion().LessThan(constants.V1_16) {
-			return errors.New("--kustomize-dir can't be used with kubeadm older than v1.16")
+		// fail fast if required to use kustomize and kubeadm less than v1.17
+		if kustomizeDir != "" && n.MustKubeadmVersion().LessThan(constants.V1_17) {
+			return errors.New("--kustomize-dir can't be used with kubeadm older than v1.17")
```
Comment (neolit123) [inline kinder/pkg/cluster/manager/actions/kubeadm-upgrade.go:50]: ditto.

File: kinder/pkg/cluster/manager/actions/kubeadm-init.go (hunk 1 above)
Comment (Akhilesh2412) [reply kinder/pkg/cluster/manager/actions/kubeadm-init.go]: Done.

File: kinder/pkg/cluster/manager/actions/kubeadm-join.go (hunk 2 above)
Comment (Akhilesh2412) [reply kinder/pkg/cluster/manager/actions/kubeadm-join.go]: Done

File: kinder/pkg/cluster/manager/actions/kubeadm-upgrade.go (hunk 3 above)
Comment (Akhilesh2412) [reply kinder/pkg/cluster/manager/actions/kubeadm-upgrade.go]: Done.

**Comment by Akhilesh2412** [comment]
@neolit123 Given the the issue is solves in #2248 do we need this PR anymore? Let's close if not needed.

**Comment by neolit123** [comment]
> @neolit123 Given the the issue is solves in #2248 do we need this PR anymore? Let's close if not needed.

i retitled that PR it had 1.16 in the title but it was about 1.15.

in kinder, 1.16 will still be used until 1.20 is released. that is because we have to maintain 1.16 -> 1.17 upgrade and skew jobs and 1.17 will be in support until 1.20 is released.

we still have to remove some of the workflows as per the "Cleanup E2E jobs for 1.x-3 (requires two PRs)" task.

- please retitle this PR to "kinder: remove 1.16 workflows"
- leave the logic around constants / checks untouched (this will be removed once we remove the kustomize functionality, i will send a PR for that at some point)
- in this PR remove the following files:
https://github.com/kubernetes/kubeadm/blob/master/kinder/ci/workflows/discovery-1.16.yaml
https://github.com/kubernetes/kubeadm/blob/master/kinder/ci/workflows/external-etcd-1.16.yaml
https://github.com/kubernetes/kubeadm/blob/master/kinder/ci/workflows/kustomize-1.16.yaml
https://github.com/kubernetes/kubeadm/blob/master/kinder/ci/workflows/regular-1.16.yaml
(notice we are keeping the upgrade and skew jobs for 1.16)

remove the same jobs from these files with a PR for test-infra:
https://github.com/kubernetes/test-infra/tree/master/config/jobs/kubernetes/sig-cluster-lifecycle

LMK if you have more questions.

**Comment by neolit123** [comment]
thanks for the update, please send the PR for https://github.com/kubernetes/test-infra/tree/master/config/jobs/kubernetes/sig-cluster-lifecycle and we can then merge that one first and this PR after it.

**Comment by Akhilesh2412** [comment]
> thanks for the update, please send the PR for https://github.com/kubernetes/test-infra/tree/master/config/jobs/kubernetes/sig-cluster-lifecycle and we can then merge that one first and this PR after it.

Here is the CL: https://github.com/kubernetes/test-infra/pull/19164 
Please review and let me know for the changes if any.

**Comment by neolit123** [comment]
the test infra PR merged.
/lgtm

**Comment by Akhilesh2412** [comment]
@neolit123 @fabriziopandini Thank you very much for the help!!! :) :)

**Pull request merged**